util/merkle/src/triedbmut.rs

```diff
@@ -0,0 +1,539 @@
+                                Some(self.insert_aux(n_path.mid(1), v_, key[n_path.at(0) as usize], old_val)?);
+
+                            let n_branch = RlpNode::Branch(partial, key);
+                            let node_rlp = RlpNode::encoded(Some(n_branch), None);
+                            let hash = self.db.insert(&node_rlp[..]);
+                            self.db.remove(&cur_node_hash.expect("Can not insert leaf into branch"));
+
+                            Ok(hash)
+                        }
+                    }
+                    None => {
+                        let node = RlpNode::Leaf(path, v_);
+                        let node_rlp = RlpNode::encoded(Some(node), None);
+                        let hash = self.db.insert(&node_rlp[..]);
+
+                        self.db.remove(&cur_node_hash.expect("Can not remove root and add new leaf"));
+
+                        Ok(hash)
+                    }
+                }
+            }
+            None => {
+                let node = RlpNode::Leaf(path, v_);
+                let node_rlp = RlpNode::encoded(Some(node), None);
+                let hash = self.db.insert(&node_rlp[..]);
+
+                Ok(hash)
+            }
+        }
+    }
+
+    /// Get auxiliary
+    fn get_aux(&self, path: NibbleSlice, cur_node_hash: Option<H256>) -> super::Result<Option<DBValue>> {
+        match cur_node_hash {
+            Some(hash) => {
+                let node_rlp = self.db.get(&hash).ok_or_else(|| Box::new(TrieError::IncompleteDatabase(hash)))?;
+
+                match RlpNode::decoded(&node_rlp[..]) {
+                    Some(RlpNode::Leaf(partial, v)) => {
+                        if partial == path {
+                            Ok(Some(v))
+                        } else {
+                            Ok(None)
+                        }
+                    }
+                    Some(RlpNode::Branch(partial, key)) => {
+                        if path.starts_with(&partial) {
+                            self.get_aux(path.mid(partial.len() + 1), key[path.mid(partial.len()).at(0) as usize])
+                        } else {
+                            Ok(None)
+                        }
+                    }
+                    None => Ok(None),
+                }
+            }
+            None => Ok(None),
+        }
+    }
+
+    /// Remove auxiliary
+    fn remove_aux(
+        &mut self,
+        path: NibbleSlice,
+        cur_node_hash: Option<H256>,
+        old_val: &mut Option<DBValue>,
+    ) -> super::Result<Option<H256>> {
+        match cur_node_hash {
+            Some(hash) => {
+                let node_rlp = self.db.get(&hash).ok_or_else(|| Box::new(TrieError::IncompleteDatabase(hash)))?;
+
+                match RlpNode::decoded(&node_rlp[..]) {
+                    Some(RlpNode::Leaf(partial, v)) => {
+                        if path == partial {
+                            self.db.remove(&cur_node_hash.expect("Can not remove leaf"));
```

Comment:
You MUST NOT remove any item in `db`.
The `remove` function just change the shape of trie.

Reply:
I remove all `db.remove()` however I wonder why there is `db.remove()` in util/patricia_trie/triedbmut?